src/Microsoft.OData.Core/UriParser/Parsers/CountSegmentParser.cs

```diff
@@ -0,0 +1,138 @@
+        /// </summary>
+        public UriQueryExpressionParser UriQueryExpressionParser
+        {
+            get { return this.parser; }
+        }
+
+        /// <summary>
+        /// Reference to the lexer.
+        /// </summary>
+        public ExpressionLexer Lexer
+        {
+            get { return this.lexer; }
+        }
+
+        /// <summary>
+        /// Create a <see cref="CountSegmentToken"/>.
+        /// </summary>
+        /// <param name="identifier">The name of this token, which in this case is always "$count".The name of this token, which in this case is always "$count".</param>
+        /// <param name="countedInstance">The instance to count on.</param>
+        public CountSegmentToken CreateCountSegmentToken(string identifier, QueryToken countedInstance)
+        {
+            QueryToken filterToken = null;
+            QueryToken searchToken = null;
+            string textWithinParenthesis = null;
+
+            // We need to maintain the recursionDepth of the outer $filter query since calling
+            // ParseFilter or ParseSearch below will reset it to 0.
+            int outerRecursiveDepth = this.UriQueryExpressionParser.recursionDepth;
+            ExpressionLexer outerLexer = this.UriQueryExpressionParser.Lexer;
+
+            if (this.lexer.CurrentToken.Kind != ExpressionTokenKind.OpenParen)
+            {
+                return new CountSegmentToken(identifier, countedInstance);
+            }
+            else
+            {
+                textWithinParenthesis = this.lexer.AdvanceThroughExpandOption();
+            }
+
+            // TODO: check for NoDollarQueryOptions
+            if (textWithinParenthesis.StartsWith(ExpressionConstants.QueryOptionFilter, StringComparison.OrdinalIgnoreCase))
+            {
+                string filterQuery = TryGetQueryOption(ExpressionConstants.QueryOptionFilter, textWithinParenthesis);
+                filterToken = filterQuery == null ? null: this.UriQueryExpressionParser.ParseFilter(filterQuery);
+            }
+            else if (textWithinParenthesis.StartsWith(ExpressionConstants.QueryOptionSearch, StringComparison.OrdinalIgnoreCase))
+            {
+                string searchQuery = TryGetQueryOption(ExpressionConstants.QueryOptionSearch, textWithinParenthesis);
+                searchToken = searchQuery == null ? null : this.UriQueryExpressionParser.ParseSearch(searchQuery);
+            }
+            else
+            {
+                // Only a $filter and $search can be inside a $count()
+                throw ParseError(ODataErrorStrings.UriQueryExpressionParser_IllegalQueryOptioninDollarCount());
+            }
+
+            if (this.lexer.CurrentToken.Kind != ExpressionTokenKind.CloseParen)
+            {
+                throw ParseError(ODataErrorStrings.UriQueryExpressionParser_CloseParenExpected(this.lexer.CurrentToken.Position, this.lexer.ExpressionText));
+            }

```

Comment:
what if /$count includes *both* $filter and $search?

Reply:
Resolved